File: fibheap.py
from collections import defaultdict


class FibonacciHeap:
    def __init__(self):
        self.min_node = None
        self.size = 0
        self.node_map = {}
        self.root_list = LL()

    def push(self, elem, key=None):
        assert elem is not None
        assert elem not in self

        self.size += 1
        new_node = self.root_list.push(elem, key)
        self.node_map[elem] = new_node
        
        if len(self) == 1:
            self.min_node = new_node
        else:
            self.min_node = min(self.min_node, new_node)

    def pop(self):
        assert not self.empty()

        elem = self.root()
        self.size -= 1
        del self.node_map[elem]

        node = self.root_list.remove(self.min_node)
        children = node.data.child_list

        ptr = children.head
        while ptr is not None:
            ptr.data.parent = None
            ptr = ptr.next

        self.root_list.merge(children)
        self._consolidate()
        self.min_node = self.root_list.min()

        return elem

    def _consolidate(self):
        rank_map = defaultdict(lambda: None)

        ptr = self.root_list.head
        while ptr is not None:
            ptr_rank = ptr.data.rank()

            if rank_map[ptr_rank] is None:
                rank_map[ptr_rank] = ptr
                ptr = ptr.next
            elif rank_map[ptr_rank] == ptr:
                ptr = ptr.next
            else:
                other_node = rank_map[ptr_rank]
                main_node, child_node = self._get_nodes(ptr, other_node)
                self.root_list.remove(child_node)
                child_node.data.parent = main_node
                main_node.data.child_list.push_node(child_node)

                rank_map[ptr_rank] = None
                ptr = self.root_list.head

        return

    def _get_nodes(self, ptr, other_node):
        if ptr.data.key <= other_node.data.key:
            return ptr, other_node
        else:
            return other_node, ptr

    def _cut(self, node):
        assert isinstance(node, LL._LL_Node)

        parent = node.data.parent
        if parent is None:
            parent.data.unmark()
            return

        parent.data.child_list.remove(node)
        node.data.parent = None
        self.root_list.push_node(node)
        node.data.unmark()

    def decrease_key(self, elem, new_key):
        assert elem in self

        node = self.node_map[elem]
        if new_key == node.data.key:
            return
        
        assert new_key < node.data.key

        node.data.key = new_key
        parent = node.data.parent
        if parent is None:
            self.min_node = min(self.min_node, node)
            return

        if parent.data.key <= new_key:
            return

        self._cut(node)
        self.min_node = min(self.min_node, node)

        node = parent
        parent = parent.data.parent
        while parent is not None:
            if not node.data.is_marked():
                node.data.mark()
                break
            else:
                self._cut(node)

                node = parent
                parent = parent.data.parent

    def root(self):
        assert not self.empty()
        return self.min_node.data.elem

    def empty(self):
        return self.__len__() == 0

    def __len__(self):
        return self.size

    def __contains__(self, elem):
        return elem in self.node_map

class LL:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def push(self, elem, key=None):
        data = FibNode(elem, key)
        node = LL._LL_Node(data)

        if len(self) == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node

        self.size += 1
        return node

    def push_node(self, node):
        assert isinstance(node, LL._LL_Node)

        if len(self) == 0:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node

        self.size += 1
        return node

    def pop(self):
        assert len(self) > 0

        node = self.tail

        if len(self) == 1:
            self.head = None
            self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None

        self.size -= 1
        node.prev = None
        node.next = None
        return node

    def remove(self, node):
        assert isinstance(node, LL._LL_Node)

        if node == self.tail:
            return self.pop()

        self.size -= 1

        if node == self.head:
            self.head = self.head.next
            self.head.prev = None

            node.prev = None
            node.next = None
            return node

        front = node.next
        back = node.prev

        back.next = front
        front.prev = back

        node.next = None
        node.prev = None
        return node

    def merge(self, other):
        assert isinstance(other, LL)

        if len(self) == 0:
            self.head = other.head
            self.tail = other.tail
            self.size = len(other)
        else:
            self.size += len(other)
            self.tail.next = other.head

            if other.head is not None:
                other.head.prev = self.tail

            if other.tail is not None:
                self.tail = other.tail

    def min(self):
        min_node = self.head
        ptr = self.head

        while ptr is not None:
            min_node = min(min_node, ptr)
            ptr = ptr.next

        return min_node

    def __len__(self):
        return self.size

    class _LL_Node:
        def __init__(self, data):
            self.next = None
            self.prev = None
            self.data = data

        def __lt__(self, other):
            return self.data < other.data

        def __gt__(self, other):
            return self.data > other.data

        def __eq__(self, other):
            return self.data == other.data

class FibNode:
    def __init__(self, elem, key):
        self.elem = elem
        self.key = key if key is not None else elem
        self.parent = None
        self.child_list = LL()
        self.marked = False

    def rank(self):
        return len(self.child_list)

    def mark(self):
        self.marked = True

    def unmark(self):
        self.marked = False

    def is_marked(self):
        return self.marked

    def __lt__(self, other):
        return self.key < other.key

    def __gt__(self, other):
        return self.key > other.key

    def __eq__(self, other):
        return self.key == other.key and self.elem == other.elem

File: C++/fibheap.cpp
#include "fibheap.h"

#include <unordered_map>
#include <assert.h>

typedef struct FibNode {
    std::string elem;
    float key;
    FibNode * parent;
    FibNode * next;
    FibNode * prev;
    FibNode * childHead;
    FibNode * childTail;
    size_t rank;
    bool marked;
} FibNode;

struct FibonacciHeap::ClassVars {
    size_t size;
    bool reverse;
    FibNode * minNode;
    FibNode * rootHead;
    FibNode * rootTail;
    std::unordered_map<std::string, FibNode *> nodeMap;
};

FibonacciHeap::FibonacciHeap(bool reverse) {
    this->ptr = new ClassVars;
    this->ptr->size = 0;
    this->ptr->reverse = reverse;
    this->ptr->minNode = NULL;
    this->ptr->rootHead = NULL;
    this->ptr->rootTail = NULL;
    this->ptr->nodeMap = std::unordered_map<std::string, FibNode *>();
}

void FibonacciHeap::copy(const FibonacciHeap & other) {
    this->ptr = new ClassVars;
    this->ptr->size = 0;
    this->ptr->reverse = other.ptr->reverse;
    this->ptr->minNode = NULL;
    this->ptr->rootHead = NULL;
    this->ptr->rootTail = NULL;
    this->ptr->nodeMap = std::unordered_map<std::string, FibNode *>();

    for (std::pair<std::string, FibNode *> kvPair : other.ptr->nodeMap) {
        std::string elem = kvPair.first;
        float key = kvPair.second->key;
        this->push(elem, key);
    }

    for (std::pair<std::string, FibNode *> kvPair : this->ptr->nodeMap) {
        std::string elem = kvPair.first;
        FibNode * node = kvPair.second;
        FibNode * otherNode = other.ptr->nodeMap[elem];

        FibNode * parent = otherNode->parent;
        FibNode * next = otherNode->next;
        FibNode * prev = otherNode->prev;
        FibNode * childHead = otherNode->childHead;
        FibNode * childTail = otherNode->childTail;

        if (parent != NULL) node->parent = this->ptr->nodeMap[parent->elem];
        if (next != NULL) node->next = this->ptr->nodeMap[next->elem];
        if (prev != NULL) node->prev = this->ptr->nodeMap[prev->elem];
        if (childHead != NULL) node->childHead = this->ptr->nodeMap[childHead->elem];
        if (childTail != NULL) node->childTail = this->ptr->nodeMap[childTail->elem];

        node->rank = otherNode->rank;
        node->marked = otherNode->marked;
    }
}

FibonacciHeap::FibonacciHeap(const FibonacciHeap & other) {
    copy(other);
}

FibonacciHeap & FibonacciHeap::operator=(const FibonacciHeap & other) {
    if (this == &other) return *this;
    
    this->~FibonacciHeap();
    copy(other);
    return *this;
}

FibonacciHeap::~FibonacciHeap() {
    for (std::pair<std::string, FibNode *> kvPair : this->ptr->nodeMap) {
        delete kvPair.second;
    }

    delete this->ptr;
}

bool FibonacciHeap::contains(std::string elem) {
    return this->ptr->nodeMap.find(elem) != this->ptr->nodeMap.end();
}

FibNode * createFibNode(std::string elem, float key) {
    FibNode * node = new FibNode;
    node->elem = elem;
    node->key = key;
    node->parent = NULL;
    node->prev = NULL;
    node->next = NULL;
    node->childHead = NULL;
    node->childTail = NULL;
    node->rank = 0;
    node->marked = false;

    return node;
}

FibNode * getMinNode(FibNode *& a, FibNode *& b, bool reverse) {
    if (reverse) {
        return a->key < b->key ? b : a;
    }

    return a->key < b->key ? a : b;
}

void FibonacciHeap::push(std::string elem, float key) {
    assert(!contains(elem));

    FibNode * node = createFibNode(elem, key);
    this->ptr->nodeMap[elem] = node;
    this->ptr->size++;
    
    if (this->ptr->size == 1) {
        this->ptr->rootHead = node;
        this->ptr->rootTail = node;
        this->ptr->minNode = node;
	    return;
    }

    this->ptr->rootTail->next = node;
    node->prev = this->ptr->rootTail;
    this->ptr->rootTail = node;

    this->ptr->minNode = getMinNode(this->ptr->minNode, node, this->ptr->reverse);
}

std::string FibonacciHeap::top() {
    assert(size() > 0);

    return this->ptr->minNode->elem;
}

size_t FibonacciHeap::size() {
    return this->ptr->size;
}

bool FibonacciHeap::empty() {
    return this->ptr->size == 0;
}

FibNode * findMinNode(FibNode *& head, bool reverse) {
    FibNode * ptr = head;
    FibNode * minNode = head;

    while (ptr != NULL) {
        minNode = getMinNode(minNode, ptr, reverse);
        ptr = ptr->next;
    }

    return minNode;
}

std::pair<FibNode *, FibNode *> getNodes(FibNode *& a, FibNode *& b, bool reverse) {
    std::pair<FibNode *, FibNode *> p(a, b);
    std::pair<FibNode *, FibNode *> p2(b, a);
    
    if (reverse) {
        return a->key < b->key ? p2 : p;
    }

    return a->key < b->key ? p : p2;
}

std::pair<FibNode *, FibNode *> removeNode(FibNode *& node, FibNode *& head, FibNode *& tail) {
    if (node == NULL) return {head, tail};

    FibNode * back = node->prev;
    FibNode * front = node->next;
    node->next = NULL;
    node->prev = NULL;

    if (back != NULL && front != NULL) {
        back->next = front;
        front->prev = back;
        return {head, tail};
    } else if (back == NULL && front == NULL) {
        return {NULL, NULL};
    } else if (back == NULL && front != NULL) {
        front->prev = NULL;
        return {front, tail};
    } else {
        back->next = NULL;
        return {head, back};
    }
}

void attachChildNode(FibNode *& mainNode, FibNode *& childNode) {
    childNode->parent = mainNode;

    if (mainNode->rank == 0) {
        mainNode->childHead = childNode;
        mainNode->childTail = childNode;
    } else {
        mainNode->childTail->next = childNode;
        childNode->prev = mainNode->childTail;
        mainNode->childTail = childNode;
    }

    mainNode->rank++;
}

std::pair<FibNode *, FibNode *> bumpNodeChildren(FibNode *& node, FibNode *& head, FibNode *& tail) {
    assert(node != NULL);

    FibNode * childHead = node->childHead;
    FibNode * childTail = node->childTail;

    FibNode * parent = head != NULL ? head->parent : NULL;
    FibNode * ptr = childHead;
    while (ptr != NULL) {
        ptr->parent = parent;
        ptr = ptr->next;
    }

    if (head == NULL) return {childHead, childTail};
    if (childHead == NULL) return {head, tail};

    tail->next = childHead;
    childHead->prev = tail;

    return {head, childTail};
}

void FibonacciHeap::consolidate() {
    std::unordered_map<size_t, FibNode *> rankMap;
    FibNode * ptr = this->ptr->rootHead;
    bool reverse = this->ptr->reverse;

    while (ptr != NULL) {
        size_t rank = ptr->rank;
        if (rankMap.find(rank) == rankMap.end() || rankMap[rank] == NULL) {
            rankMap[rank] = ptr;
            ptr = ptr->next;
        } else if (rankMap[rank] == ptr) {
            ptr = ptr->next;
        } else {
            FibNode * otherNode = rankMap[rank];
            std::pair<FibNode *, FibNode *> nodes = getNodes(ptr, otherNode, reverse);
            FibNode * mainNode = nodes.first;
            FibNode * childNode = nodes.second;
            
            std::pair<FibNode *, FibNode *> p = removeNode(childNode, this->ptr->rootHead, this->ptr->rootTail);
            this->ptr->rootHead = p.first;
            this->ptr->rootTail = p.second;

            attachChildNode(mainNode, childNode);

            rankMap[rank] = NULL;
            ptr = this->ptr->rootHead;
        }
    }
}

void FibonacciHeap::pop() {
    assert(!empty());

    std::string elem = top();
    this->ptr->nodeMap.erase(elem);
    this->ptr->size--;

    std::pair<FibNode *, FibNode *> p = removeNode(this->ptr->minNode, this->ptr->rootHead, this->ptr->rootTail);
    this->ptr->rootHead = p.first;
    this->ptr->rootTail = p.second;

    FibNode * minNode = this->ptr->minNode;
    p = bumpNodeChildren(minNode, this->ptr->rootHead, this->ptr->rootTail);
    this->ptr->rootHead = p.first;
    this->ptr->rootTail = p.second;
    delete minNode;

    consolidate();
    this->ptr->minNode = findMinNode(this->ptr->rootHead, this->ptr->reverse);
}

void cut(FibNode *& node, FibNode *& parent) {
    if (parent == NULL) {
        parent->marked = false;
        return;
    }

    // remove node from parent's child list
    std::pair<FibNode *, FibNode *> p = removeNode(node, parent->childHead, parent->childTail);
    parent->childHead = p.first;
    parent->childTail = p.second;

    parent->rank--;
}

void FibonacciHeap::decreaseKey(std::string elem, float newKey) {
    assert(contains(elem));

    FibNode * node = this->ptr->nodeMap[elem];
    if (newKey == node->key) return;
    bool reverse = this->ptr->reverse;

    assert(
        (!reverse && newKey < node->key) ||
        (reverse && newKey > node->key)
    );

    node->key = newKey;
    FibNode * parent = node->parent;

    if (parent == NULL) {
        this->ptr->minNode = getMinNode(this->ptr->minNode, node, reverse);
        return;
    }

    if (!reverse && parent->key <= newKey) return;
    if (reverse && parent->key >= newKey) return;

    cut(node, parent);
    node->parent = NULL;
    this->ptr->rootTail->next = node;
    node->prev = this->ptr->rootTail;
    this->ptr->rootTail = node;
    node->marked = false;

    this->ptr->minNode = getMinNode(this->ptr->minNode, node, reverse); 

    node = parent;
    parent = parent->parent;
    while (parent != NULL) {
        if (!node->marked) {
            node->marked = true;
            break;
        }

        cut(node, parent);
        node->parent = NULL;
        this->ptr->rootTail->next = node;
        node->prev = this->ptr->rootTail;
        this->ptr->rootTail = node;
        node->marked = false;

        node = parent;
        parent = parent->parent;
    }
}
